ToDoList-Redux/AppLifecycle.swift

```diff
@@ -0,0 +1,84 @@
+//
+//  AppLifecycleMiddleware.swift
+//  ToDoList-Redux
+//
+//  Created by Luiz Rodrigo Martins Barbosa on 11.08.20.
+//
+
+import Combine
+import Foundation
+import SwiftRex
+import UIKit
+
+enum AppLifecycleAction {
+    case didEnterBackground
+    case willEnterForeground
+    case didBecomeActive
+    case willBecomeInactive
+}
+
+enum AppLifecycle: Equatable {
+    case backgroundActive
+    case backgroundInactive
+    case foregroundActive
+    case foregroundInactive
+}
+
+extension Reducer where ActionType == AppLifecycleAction, StateType == AppLifecycle {
+    static let lifecycle = Reducer { action, state in
+        switch (state, action) {
+        case (.backgroundActive, .didEnterBackground): return state
+        case (.backgroundInactive, .didEnterBackground): return state
+        case (.foregroundActive, .didEnterBackground): return .backgroundActive
+        case (.foregroundInactive, .didEnterBackground): return .backgroundInactive
+
+        case (.backgroundActive, .willEnterForeground): return .foregroundActive
+        case (.backgroundInactive, .willEnterForeground): return .foregroundInactive
+        case (.foregroundActive, .willEnterForeground): return state
+        case (.foregroundInactive, .willEnterForeground): return state
+
+        case (.backgroundActive, .didBecomeActive): return state
+        case (.backgroundInactive, .didBecomeActive): return .backgroundActive
+        case (.foregroundActive, .didBecomeActive): return state
+        case (.foregroundInactive, .didBecomeActive): return .foregroundActive
+
+        case (.backgroundActive, .willBecomeInactive): return .backgroundInactive
+        case (.backgroundInactive, .willBecomeInactive): return state
+        case (.foregroundActive, .willBecomeInactive): return .foregroundInactive
+        case (.foregroundInactive, .willBecomeInactive): return state
+        }
+    }
+}
+
+// TODO: Later change to upcoming (0.8) EffectMiddleware
+class AppLifecycleMiddleware: Middleware {
```

Comment:
This file is just a basic implementation of App Lifecycle, you have everything you need here, and you can even use in different apps because it's totally decoupled from your AppState/AppAction. One thing to notice: perhaps the initial state will be wrong in case the Middleware is created after the app already changed to Active or Foreground, or when app is coming from deep-linking, APNS, background mode, etc. Please use this carefully and be sure to test these scenarios in a real-life app.